cli/pkg/cmdutil/archive.go

```diff
@@ -0,0 +1,124 @@
+package cmdutil
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	adminv1 "github.com/rilldata/rill/proto/gen/rill/admin/v1"
+	"github.com/rilldata/rill/runtime/drivers"
+)
+
+// UploadRepo uploads a local project artifacts to rill managed store and returns upload path
+func UploadRepo(ctx context.Context, repo drivers.RepoStore, ch *Helper, org, project string) (string, error) {
+	// list files
+	entries, err := repo.ListRecursive(ctx, "**", false)
+	if err != nil {
+		return "", err
+	}
+
+	// generate a tar ball
+	b := &bytes.Buffer{}
+	if err := createTarball(b, entries, repo.Root()); err != nil {
+		return "", err
+	}
+
+	// upload the tar ball
+	uploadPath, err := uploadTarBall(ctx, ch, org, project, b)
+	if err != nil {
+		return "", err
+	}
+	return uploadPath, nil
+}
+
+// borrowed from https://github.com/goreleaser/goreleaser/blob/main/pkg/archive/tar/tar.go with minor changes
+func createTarball(writer io.Writer, files []drivers.DirEntry, root string) error {
+	gw, err := gzip.NewWriterLevel(writer, gzip.BestCompression)
+	if err != nil {
+		return err
+	}
+	defer gw.Close()
+	tw := tar.NewWriter(gw)
+	defer tw.Close()
+	for _, entry := range files {
+		if strings.EqualFold(entry.Path, "/.env") { // ignore .env
+			continue
+		}
+		fullPath := filepath.Join(root, entry.Path)
+		info, err := os.Lstat(fullPath)
+		if err != nil {
+			return fmt.Errorf("%s: %w", fullPath, err)
+		}
+		if info.Mode()&os.ModeSymlink != 0 {
+			return fmt.Errorf("%s: repo contains symlinks", entry.Path)
+		}
+
+		header, err := tar.FileInfoHeader(info, "")
+		if err != nil {
+			return fmt.Errorf("%s: %w", fullPath, err)
+		}
+		header.Name = entry.Path
+		if err = tw.WriteHeader(header); err != nil {
+			return fmt.Errorf("%s: %w", fullPath, err)
+		}
+		if info.IsDir() {
+			continue
+		}
+
+		file, err := os.Open(fullPath)
+		if err != nil {
+			return fmt.Errorf("%s: %w", fullPath, err)
+		}
+		if _, err := io.Copy(tw, file); err != nil {
+			file.Close()
+			return fmt.Errorf("%s: %w", fullPath, err)
+		}
+		file.Close()
+	}
+	return nil
+}
+
+func uploadTarBall(ctx context.Context, ch *Helper, org, project string, body io.Reader) (string, error) {
+	adminClient, err := ch.Client()
+	if err != nil {
+		return "", err
+	}
+
+	// generate a upload URL
+	uploadURL, err := adminClient.CreateUploadSignedURL(ctx, &adminv1.CreateUploadSignedURLRequest{
+		OrganizationName: org,
+		ProjectName:      project,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	// Create a put request
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, uploadURL.SignedUrl, body)
+	if err != nil {
+		return "", fmt.Errorf("failed to create request: %w", err)
+	}
+	req.Header.Set("Content-Type", "application/octet-stream")
+	req.Header.Set("x-goog-content-length-range", "1,104857600")
```

Comment:
If this is necessary, I think the headers to use should be returned from `CreateUploadSignedURL` to avoid hard-coding this in two places

Reply:
Yes this is necessary. The headers should match else we get signature errors.
Okay we can headers in the API response.